unsafe blocks: add options to Makefile and new script to print stats

stefano-garzarella opened the pull request:
This PR adds `UNSAFE_BLOCKS` and `CLIPPY_OPTIONS` to the Makefile.
These are used by the new `unsafe_blocks_report.sh` also added by this PR to print stats about undocumented unsafe blocks in our base code.

I'll open a new PR when this is merged to use the script in the CI, since our pipeline requires to have the changes in the base branch

joergroedel commented on Makefile:
```diff
@@ -9,6 +9,12 @@ ifneq ($(FEATURES_TEST),)
 
+CLIPPY_OPTIONS ?=
+CLIPPY_ARGS ?= -D warnings
+ifdef UNSAFE_BLOCKS
+CLIPPY_ARGS = -W clippy::undocumented_unsafe_blocks
+endif
```
This removes any external `CLIPPY_ARGS`, no? Better append the new `-W` parameter to the existing ones.

stefano-garzarella replied:
It's on purpose as I wrote in the commit description:

> `make clippy UNSAFE_BLOCKS=1` will enable
`clippy::undocumented_unsafe_blocks` warnings. This can help the effort
to reduce them. If it's set we remove `-D warnings` since `clippy`
will exit after checking the first module that contains them.

stefano-garzarella replied:
Maybe I can check if I can override `-D warnings`

stefano-garzarella replied:
```diff
diff --git a/Makefile b/Makefile
index 8a38fe2c..79dcaa8f 100644
--- a/Makefile
+++ b/Makefile
@@ -12,7 +12,7 @@ endif
 CLIPPY_OPTIONS ?=
 CLIPPY_ARGS ?= -D warnings
 ifdef UNSAFE_BLOCKS
-CLIPPY_ARGS = -W clippy::undocumented_unsafe_blocks
+CLIPPY_ARGS += -W warnings -W clippy::undocumented_unsafe_blocks
 endif
 
 ifdef RELEASE
```

This seams to have the same output, @joergroedel  WDYT?
(`-W warnings` will re-enable the warnings to be considered as warnings and not errors)

stefano-garzarella replied:
Done and updated the commit description ;-)

joergroedel reviewed (commented):
LGTM, except the one question on the Makefile.

joergroedel reviewed (dismissed):
LGTM, thanks!

luigix25 commented:
LGTM!

stefano-garzarella commented:
v3:
- removed first `/` on file's path [@joergroedel @peterfang]
- rebased on main (commit c5a82785bc643732db247e1c41f0ef8572e1f050)

This is the current output:
```
$ ./scripts/unsafe_blocks_report.sh  
Running `make clippy`...
make: Entering directory '/home/stefano/repos/coconut/svsm-work'
cargo clippy --quiet --color=never --all-features --workspace --exclude svsm --exclude stage1 --exclude svsm-fuzz -- -D warnings -W warnings -W clippy::undocumented_unsafe_blocks
RUSTFLAGS="--cfg fuzzing" cargo clippy --quiet --color=never --all-features --package svsm-fuzz -- -D warnings -W warnings -W clippy::undocumented_unsafe_blocks
cargo clippy --quiet --color=never --all-features --package svsm --target x86_64-unknown-none -- -D warnings -W warnings -W clippy::undocumented_unsafe_blocks
cargo clippy --quiet --color=never --all-features --package stage1 --target x86_64-unknown-none -- -D warnings -W warnings -W clippy::undocumented_unsafe_blocks -C panic=abort
cargo clippy --quiet --color=never --all-features --workspace --tests --exclude packit -- -D warnings -W warnings -W clippy::undocumented_unsafe_blocks
make: Leaving directory '/home/stefano/repos/coconut/svsm-work'

Undocumented unsafe blocks per module

    119 kernel
    119 kernel/src
     30 kernel/src/cpu
      3 kernel/src/cpu/ipi.rs
      2 kernel/src/cpu/percpu.rs
      1 kernel/src/cpu/shadow_stack.rs
      1 kernel/src/cpu/sse.rs
     21 kernel/src/cpu/vc.rs
      2 kernel/src/cpu/vmsa.rs
      9 kernel/src/debug
      5 kernel/src/debug/gdbstub.rs
      4 kernel/src/debug/stacktrace.rs
     10 kernel/src/insn_decode
     10 kernel/src/insn_decode/insn.rs
      4 kernel/src/locking
      2 kernel/src/locking/rwlock.rs
      2 kernel/src/locking/spinlock.rs
     62 kernel/src/mm
     21 kernel/src/mm/alloc.rs
     13 kernel/src/mm/guestmem.rs
      9 kernel/src/mm/pagebox.rs
      7 kernel/src/mm/pagetable.rs
      5 kernel/src/mm/page_visibility.rs
      4 kernel/src/mm/ptguards.rs
      1 kernel/src/mm/validate.rs
      2 kernel/src/mm/vm
      2 kernel/src/mm/vm/mapping
      2 kernel/src/mm/vm/mapping/api.rs
      4 kernel/src/task
      2 kernel/src/task/schedule.rs
      2 kernel/src/task/tasks.rs
     20 virtio-drivers
     20 virtio-drivers/src
      6 virtio-drivers/src/hal
      6 virtio-drivers/src/hal/fake.rs
     14 virtio-drivers/src/queue.rs

Total number of undocumented unsafe blocks

139
```
